### ISSUE TITLE
Conduit proxy in controller pod can't connect with proxy api Service DNS

### ISSUE THREAD

**Issue by deebo91** [runconduit/conduit#302]
This issue arises from #227. Initially, we try to inject a proxy into the controller pod to ensure that the proxy routes traffic within the pod as well as to other services outside the pod. When making requests curl requests from within the pod to other services like prometheus, the proxy is unable to route the request and just hangs. 

Other things we see in this setup is that conduit proxy container cannot connect to the proxy api through the Proxy API Kubernetes DNS name set by `CONDUIT_PROXY_CONTROL_URL` environment variable. The variable is the service DNS for the proxy api. Changing the environment variable to `tcp://localhost:8086` in the conduit proxy container lets the conduit proxy connect successfully to the proxy api and routes traffic successfully from within the controller pod.

@briansmith, you may have more insight as to what is going on. Feel free to modify or add to the description of the issue or make any corrections.

**Comment by briansmith** [comment]
@deebo91 Are you saying that setting CONDUIT_PROXY_CONTROL_URL=tcp://localhost:8086 for the controller pod (only) appears to fix everything? I think it's probably good to override the control plane URL with a localhost one anyway, to avoid chicken-and-egg problems with the TLS bootstrapping.

> When making requests curl requests from within the pod to other services like prometheus, the proxy is unable to route the request and just hangs.

I think this sentence got mixed up during editing. Does `curl` also not work, or `curl` does work? What doesn't work?

**Comment by olix0r** [comment]
The issue I observed was the following:

I started by running:

```
bin/conduit install | bin/conduit inject - | kubectl apply -f -
```

Checking the controller's proxy, I saw:
```
:; kubectl logs -n conduit -fc conduit-proxy controller-86ff84f7d-r7dtm  
1518113800.399115 INFO conduit_proxy using controller at HostAndPort { host: Domain("proxy-api.conduit.svc.cluster.local"), port: 8086 }
1518113800.399417 INFO conduit_proxy routing on V4(127.0.0.1:4140)
1518113800.399515 INFO conduit_proxy proxying on V4(0.0.0.0:4143) to None
1518113800.402065 INFO conduit_proxy::transport::connect "controller-client", DNS resolved proxy-api.conduit.svc.cluster.local to 10.103.95.40
1518113800.408260 WARN conduit_proxy::control "controller-client", controller error: Connect(Connect(Error(Error { repr: Os { code: 111, message: "Connection refused" } })))
1518113803.766633 ERR! conduit_proxy::map_err turning service error into 500: Inner(Inner(Inner(Error { kind: Inner(Error { kind: Proto(INTERNAL_ERROR) }) })))
1518113805.411756 INFO conduit_proxy::transport::connect "controller-client", DNS resolved proxy-api.conduit.svc.cluster.local to 10.103.95.40
1518113806.203823 ERR! conduit_proxy::map_err turning service error into 500: Inner(Inner(Inner(Error { kind: Inner(Error { kind: Proto(INTERNAL_ERROR) }) })))
1518113810.403414 WARN conduit_proxy::control::telemetry "controller-client", controller error: Grpc(Status { code: Unavailable })
```

```
:; kubectl exec -it -c conduit-proxy -n conduit controller-86ff84f7d-r7dtm  bash
I have no name!@controller-86ff84f7d-r7dtm:/$ nghttp -v http://proxy-api.conduit.svc.cluster.local:8086/
[  0.012] send SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
[  0.012] send HEADERS frame <length=52, flags=0x05, stream_id=1>
          ; END_STREAM | END_HEADERS
          (padlen=0)
          ; Open new stream
          :authority: proxy-api.conduit.svc.cluster.local:8086
          :method: GET
          :path: /
          :scheme: http
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/0.6.4
[  0.018] recv SETTINGS frame <length=0, flags=0x00, stream_id=0>
          (niv=0)
[  0.018] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.020] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.026] recv (stream_id=1, noind=0) :status: 200
[  0.026] recv (stream_id=1, noind=0) content-type: application/grpc
[  0.026] recv (stream_id=1, noind=0) grpc-status: 8
[  0.026] recv (stream_id=1, noind=0) grpc-message: malformed method name: "/"
[  0.026] recv HEADERS frame <length=58, flags=0x05, stream_id=1>
          ; END_STREAM | END_HEADERS
          (padlen=0)
          ; First response header
[  0.026] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
          (last_stream_id=0, error_code=NO_ERROR(0x00), opaque_data(0)=[])
```

I don't know why the proxy is unable to connect, when it seems that other processes in the _same container_ are able to communicate.

I'm also in favor of just using localhost where we can.

@briansmith Would you mind linking/creating an issue that describes the objectives for this work?  I'm guessing we'll want to support a flag to `install` that controls whether the proxy is installed or not, but I think I'm lacking some of the context on how proxy-in-the-controller is to be bootstrapped/used.

**Comment by deebo91** [comment]
Oops, I apologize for the confusion. Yes that is correct.  `curl`ing from a container inside the controller pod doesn't work when `CONDUIT_PROXY_CONTROL_URL` is set to `proxy-api.conduit.svc.cluster.local:8086` but works when set to `tcp://localhost:8086`

**Comment by briansmith** [comment]
@deebo91 What target are you using for the curl command? That is, what is the exact command line for curl that you are using, and which container are you using curl in?

**Comment by deebo91** [comment]
I am in the conduit-proxy container, issuing a `nghttp -v http://proxy-api.conduit.svc.cluster.local:8086` Sorry, I misspoke when I said I am using curl, I am using `nghttp` to hit the endpoint since the proxy API is H2

**Comment by olix0r** [comment]
The behavior I describe a above is actually a bit more subtle...

Running with:
```
        - name: CONDUIT_PROXY_LOG
          value: tokio_core=warn,mio=warn,h2::proto=warn,debug
```

As soon as the proxy starts, it resolves the proxy-api service and starts connecting...

```
1518139819.749975 INFO conduit_proxy using controller at HostAndPort { host: Domain("proxy-api.conduit.svc.cluster.local"), port: 8086 }
1518139819.750059 INFO conduit_proxy routing on V4(127.0.0.1:4140)
1518139819.750107 INFO conduit_proxy proxying on V4(0.0.0.0:4143) to None
1518139819.751126 INFO conduit_proxy::transport::connect "controller-client", DNS resolved proxy-api.conduit.svc.cluster.local to 10.103.95.40
```

Then, immediately, we see a connection being established outbound -- presumably from one of the controller services to the k8s API:
```
1518139820.581884 DBUG conduit_proxy::transport::so_original_dst "outbound", get_original_dst TcpStream { sys: TcpStream { inner: TcpStream { addr: V4(127.0.0.1:4140), peer: V4(10.1.0.102:53404), fd: 14 } }, selector_id: SelectorId { id: AtomicUsize(1) } }
1518139820.582141 DBUG conduit_proxy::telemetry::sensor server connection open
1518139820.582272 DBUG conduit_proxy::transparency::tcp tcp accepted, forwarding (10.1.0.102:53404) to 10.96.0.1:443
```

The initial connection attempt to the controller fails:
```
1518139820.582468 DBUG conduit_proxy::telemetry::sensor::transport client connection open
1518139822.831328 WARN conduit_proxy::control "controller-client", controller error: Connect(Connect(Timeout(Duration { secs: 3, nanos: 0 })))
```


```
1518139827.842925 INFO conduit_proxy::transport::connect "controller-client", DNS resolved proxy-api.conduit.svc.cluster.local to 10.103.95.40
1518139827.843398 DBUG conduit_proxy::transport::so_original_dst "inbound", get_original_dst TcpStream { sys: TcpStream { inner: TcpStream { addr: V4(10.1.0.102:4143), peer: V4(10.1.0.1:56414), fd: 16 } }, selector_id: SelectorId { id: AtomicUsize(1) } }
1518139827.843832 DBUG h2::client "controller-client", binding client connection
1518139827.844022 DBUG h2::client "controller-client", client connection bound
1518139827.844164 DBUG h2::codec::framed_write "controller-client", send; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.844059 DBUG conduit_proxy::telemetry::sensor server connection open
1518139827.845274 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.846410 DBUG h2::codec::framed_read ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), received; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.846722 DBUG h2::codec::framed_write ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), send; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847050 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847079 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847097 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847204 DBUG h2::codec::framed_read ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), received; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.751525 DBUG h2::codec::framed_write ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), send; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
1518139829.752044 DBUG h2::codec::framed_write ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), send; frame=Frame::Data(Data { stream_id: StreamId(1), flags: DataFlags { end_stream: true }, pad_len: None })
1518139829.753446 DBUG h2::codec::framed_read received; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
1518139829.753506 DBUG h2::codec::framed_read received; frame=Frame::Data(Data { stream_id: StreamId(1), flags: DataFlags { end_stream: true }, pad_len: None })
1518139829.753525 DBUG conduit_proxy::inbound building inbound Http2 client to 10.1.0.102:8086
1518139829.753810 DBUG conduit_proxy::telemetry::sensor::transport client connection open
1518139829.753835 DBUG h2::client binding client connection
1518139829.753937 DBUG h2::client client connection bound
1518139829.753964 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: Some(0), max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.754011 DBUG h2::codec::framed_write send; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
1518139829.754148 DBUG h2::codec::framed_write send; frame=Frame::Data(Data { stream_id: StreamId(1), flags: DataFlags { end_stream: true }, pad_len: None })
1518139829.757707 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.757743 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.758376 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.758506 DBUG h2::codec::framed_read received; frame=Frame::WindowUpdate(WindowUpdate { stream_id: StreamId(0), size_increment: 136 })
1518139829.759954 DBUG h2::codec::framed_read received; frame=Frame::Ping(Ping { ack: false, payload: [2, 4, 16, 16, 9, 14, 7, 7] })
1518139829.759979 DBUG h2::codec::framed_write send; frame=Frame::Ping(Ping { ack: true, payload: [2, 4, 16, 16, 9, 14, 7, 7] })
1518139829.760433 DBUG h2::codec::framed_read received; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
1518139829.761471 DBUG h2::codec::framed_write send; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
1518139829.761906 DBUG h2::codec::framed_read ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), received; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
1518139829.762308 WARN conduit_proxy::control::telemetry "controller-client", controller error: Grpc(Status { code: Unavailable })
```

**Comment by olix0r** [comment]
The behavior I describe a above is actually a bit more subtle...

Running with:
```
        - name: CONDUIT_PROXY_LOG
          value: tokio_core=warn,mio=warn,h2::proto=warn,debug
```

As soon as the proxy starts, it resolves the proxy-api service and starts connecting...

```
1518139819.749975 INFO conduit_proxy using controller at HostAndPort { host: Domain("proxy-api.conduit.svc.cluster.local"), port: 8086 }
1518139819.750059 INFO conduit_proxy routing on V4(127.0.0.1:4140)
1518139819.750107 INFO conduit_proxy proxying on V4(0.0.0.0:4143) to None
1518139819.751126 INFO conduit_proxy::transport::connect "controller-client", DNS resolved proxy-api.conduit.svc.cluster.local to 10.103.95.40
```

Then, immediately, we see a connection being established outbound -- presumably from one of the controller services to the k8s API:
```
1518139820.581884 DBUG conduit_proxy::transport::so_original_dst "outbound", get_original_dst TcpStream { sys: TcpStream { inner: TcpStream { addr: V4(127.0.0.1:4140), peer: V4(10.1.0.102:53404), fd: 14 } }, selector_id: SelectorId { id: AtomicUsize(1) } }
1518139820.582141 DBUG conduit_proxy::telemetry::sensor server connection open
1518139820.582272 DBUG conduit_proxy::transparency::tcp tcp accepted, forwarding (10.1.0.102:53404) to 10.96.0.1:443
```

The initial connection attempt to the controller fails:
```
1518139820.582468 DBUG conduit_proxy::telemetry::sensor::transport client connection open
1518139822.831328 WARN conduit_proxy::control "controller-client", controller error: Connect(Connect(Timeout(Duration { secs: 3, nanos: 0 })))
```

I'm guessing this is because some part of the controller doesn't accept connections until
it's communicated with k8s.

But then, as soon as we retry the connection, it completes:


```
1518139827.842925 INFO conduit_proxy::transport::connect "controller-client", DNS resolved proxy-api.conduit.svc.cluster.local to 10.103.95.40
1518139827.843398 DBUG conduit_proxy::transport::so_original_dst "inbound", get_original_dst TcpStream { sys: TcpStream { inner: TcpStream { addr: V4(10.1.0.102:4143), peer: V4(10.1.0.1:56414), fd: 16 } }, selector_id: SelectorId { id: AtomicUsize(1) } }
1518139827.843832 DBUG h2::client "controller-client", binding client connection
1518139827.844022 DBUG h2::client "controller-client", client connection bound
```

And then the proxy actually connects to itself!!!

```
1518139827.844164 DBUG h2::codec::framed_write "controller-client", send; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.844059 DBUG conduit_proxy::telemetry::sensor server connection open
1518139827.845274 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.846410 DBUG h2::codec::framed_read ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), received; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.846722 DBUG h2::codec::framed_write ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), send; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847050 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847079 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847097 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139827.847204 DBUG h2::codec::framed_read ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), received; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.751525 DBUG h2::codec::framed_write ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), send; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
1518139829.752044 DBUG h2::codec::framed_write ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), send; frame=Frame::Data(Data { stream_id: StreamId(1), flags: DataFlags { end_stream: true }, pad_len: None })
1518139829.753446 DBUG h2::codec::framed_read received; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
1518139829.753506 DBUG h2::codec::framed_read received; frame=Frame::Data(Data { stream_id: StreamId(1), flags: DataFlags { end_stream: true }, pad_len: None })
```

And it routes, properly, to the process on the pod-local IP.

```
1518139829.753525 DBUG conduit_proxy::inbound building inbound Http2 client to 10.1.0.102:8086
1518139829.753810 DBUG conduit_proxy::telemetry::sensor::transport client connection open
1518139829.753835 DBUG h2::client binding client connection
1518139829.753937 DBUG h2::client client connection bound
1518139829.753964 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: Some(0), max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.754011 DBUG h2::codec::framed_write send; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: false, end_headers: true, padded: false, priority: false } })
1518139829.754148 DBUG h2::codec::framed_write send; frame=Frame::Data(Data { stream_id: StreamId(1), flags: DataFlags { end_stream: true }, pad_len: None })
1518139829.757707 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(0), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.757743 DBUG h2::codec::framed_write send; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.758376 DBUG h2::codec::framed_read received; frame=Frame::Settings(Settings { flags: SettingsFlags(1), header_table_size: None, enable_push: None, max_concurrent_streams: None, initial_window_size: None, max_frame_size: None, max_header_list_size: None })
1518139829.758506 DBUG h2::codec::framed_read received; frame=Frame::WindowUpdate(WindowUpdate { stream_id: StreamId(0), size_increment: 136 })
1518139829.759954 DBUG h2::codec::framed_read received; frame=Frame::Ping(Ping { ack: false, payload: [2, 4, 16, 16, 9, 14, 7, 7] })
1518139829.759979 DBUG h2::codec::framed_write send; frame=Frame::Ping(Ping { ack: true, payload: [2, 4, 16, 16, 9, 14, 7, 7] })
```

And then one of the peers replies with a GRPC error code.
```
1518139829.760433 DBUG h2::codec::framed_read received; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
1518139829.761471 DBUG h2::codec::framed_write send; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
1518139829.761906 DBUG h2::codec::framed_read ("controller-client", "proxy-api.conduit.svc.cluster.local:8086"), received; frame=Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } })
1518139829.762308 WARN conduit_proxy::control::telemetry "controller-client", controller error: Grpc(Status { code: Unavailable })
```

It's still not super clear which side is doing what, but this does at least confirm the
traffic comes back through the proxy. I'd want to see logging proxy-api to dig deeper into
what the error state is.

**Comment by deebo91** [comment]
I ran 
```bash
bin/conduit install | bin/conduit inject - | kubectl apply -f -
```
in Kubernetes on Docker as well as a test cluster of my own (GKE 1.7). It seems as if the proxy works fine inside the controller pod. I ran the following commands and was able to get these results.

Conduit Check command ✅ 
```bash
conduit git:(master) ✗ bin/conduit check

kubectl: is in $PATH............................................................[ok]
kubectl: has compatible version.................................................[ok]
kubectl: can talk to Kubernetes cluster.........................................[ok]
kubernetes-api: can initialize the client.......................................[ok]
kubernetes-api: can query the Kubernetes API....................................[ok]
conduit-api: can retrieve status via gRPC.......................................[ok]
conduit-api[telemetry]: control plane can use telemetry service.................[ok]

Status check results are [ok]
```
Conduit Get Pods ✅ 
```bash
conduit git:(master) ✗ bin/conduit get pods

conduit/prometheus-615540357-6hbgf
conduit/controller-1956427135-d3tv7
conduit/web-1695714391-3v7vq
```

Conduit Stat command for both deployments and pods ✅ 
```bash
conduit git:(master) ✗ conduit stat deployments

NAME                 REQUEST_RATE   SUCCESS_RATE   P50_LATENCY   P99_LATENCY
conduit/controller         0.8rps        100.00%         499ms         990ms
conduit/prometheus         0.0rps          0.00%          32ms         495ms
conduit/web                0.1rps        100.00%         750ms         995ms

conduit git:(master) ✗ conduit stat pods

NAME                                  REQUEST_RATE   SUCCESS_RATE   P50_LATENCY   P99_LATENCY
conduit/controller-1956427135-d3tv7         0.8rps        100.00%         666ms         993ms
conduit/web-1695714391-3v7vq                0.1rps        100.00%         750ms         995ms
```
shelling into the conduit proxy container and issuing an H@ get to the proxy-api returns a result and does not hang.
```bash
conduit git:(master) ✗ kubectl exec -it controller-1956427135-d3tv7 -c conduit-proxy  /bin/bash -n conduit

I have no name!@controller-6cdb577d89-qfg4w:/$ nghttp -v http://proxy-api.conduit.svc.cluster.local:8086

[  0.393] send SETTINGS frame <length=12, flags=0x00, stream_id=0>
          (niv=2)
          [SETTINGS_MAX_CONCURRENT_STREAMS(0x03):100]
          [SETTINGS_INITIAL_WINDOW_SIZE(0x04):65535]
[  0.393] send HEADERS frame <length=52, flags=0x05, stream_id=1>
          ; END_STREAM | END_HEADERS
          (padlen=0)
          ; Open new stream
          :authority: proxy-api.conduit.svc.cluster.local:8086
          :method: GET
          :path: /
          :scheme: http
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/0.6.4
[  0.425] recv SETTINGS frame <length=0, flags=0x00, stream_id=0>
          (niv=0)
[  0.425] send SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.428] recv SETTINGS frame <length=0, flags=0x01, stream_id=0>
          ; ACK
          (niv=0)
[  0.607] recv (stream_id=1, noind=0) :status: 200
[  0.607] recv (stream_id=1, noind=0) content-type: application/grpc
[  0.607] recv (stream_id=1, noind=0) grpc-status: 8
[  0.607] recv (stream_id=1, noind=0) grpc-message: malformed method name: "/"
[  0.607] recv HEADERS frame <length=58, flags=0x05, stream_id=1>
          ; END_STREAM | END_HEADERS
          (padlen=0)
          ; First response header
[  0.607] send GOAWAY frame <length=8, flags=0x00, stream_id=0>
```

I think it is safe to close this issue for now and we could reopen it if we start seeing issues again. I will leave this issue open for a little bit to wait for any comments or suggestion just in case there are more things to try.

**Comment by briansmith** [comment]
> I think it is safe to close this issue for now and we could reopen it if we start seeing issues again.

OK, closing. LMK if you are having any trouble with this part of the project.

**Comment by deebo91** [comment]
Thanks @briansmith. I have a branch https://github.com/runconduit/conduit/pull/347 that injects the controller pod with the proxy. I could use your help in verifying whether traffic is routed properly. My test was to first run the new install command. Then, install emojivoto in its own namespace. Then I verified that I could see stats for `conduit/controller` pods as well as `emojivoto/*` pods. I also took a look at the logs in the conduit proxy with `tokio_core=warn,mio=warn,h2::proto=warn,debug` set as the logging level and I am seeing h2 traffic being sent to telemetry. Does that suffice as exit criteria?
